internal/overlord/logstate/manager.go

```diff
@@ -0,0 +1,85 @@
+package logstate
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+)
+
+var logBackends = map[string]func(*plan.LogTarget) (LogBackend, error){}
+
+func RegisterLogBackend(name string, builder func(*plan.LogTarget) (LogBackend, error)) {
+	validator := func(t *plan.LogTarget) error {
+		backend, err := builder(t)
+		if backend != nil {
+			backend.Close()
+		}
+		return err
+	}
+	plan.RegisterLogBackend(name, validator)
+}
+
+type LogManager struct {
+	mutex   sync.Mutex
+	targets map[string]*LogTarget
+	// targetsByService maps service name to the list of targets to receive its logs.
+	targetsByService map[string][]*LogTarget
+	// targetSelection maps log target names to selection criteria.
+	targetSelection map[string]string
+}
+
+func NewLogManager() *LogManager {
+	return &LogManager{
+		targets:          make(map[string]*LogTarget),
+		targetsByService: make(map[string][]*LogTarget),
+		targetSelection:  make(map[string]string),
+	}
+}
+
+func (m *LogManager) GetTargets(serviceName string) ([]*LogTarget, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	targets, ok := m.targetsByService[serviceName]
+	if !ok {
+		return nil, fmt.Errorf("no known logging targets for service %q", serviceName)
+	}
+	return targets, nil
+}
+
+// PlanChanged handles updates to the plan (server configuration),
+// stopping the previous checks and starting the new ones as required.
+func (m *LogManager) PlanChanged(p *plan.Plan) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if len(p.LogTargets) == 0 {
+		return
+	}
+
+	// update targets
+	for name, target := range p.LogTargets {
+		backend, err := logBackends[name](target)
```

Comment:
I think a new method should be created that returns the static configured backends each time it's called.

This allows us to remove all of the init methods on each of the backends and the potential for loosing errors and having nasty init problems. It also would make the return static struct constant and not open for change later on.